File: game-ui-harnesses/ui-decomposition-harness/src/ai_ui_decomposition/button_label_lines.py
"""Mirror the consumer button-label-lines-v1 contract; no content inference."""
import math
from .common import require

def validate_label_lines(value,label=None,size=None):
    require(isinstance(value,dict) and set(value)=={'version','coordinateSpace','lines'} and value['version']=='1.0' and value['coordinateSpace']=='target-component-local','BUTTON_LABEL_LINES_VERSION')
    lines=value['lines'];require(isinstance(lines,list) and 1<=len(lines)<=8,'BUTTON_LABEL_LINES_COUNT')
    rects=[]
    for line in lines:
        require(isinstance(line,dict) and set(line)=={'text','fontSize','fontWeight','align','layout'} and isinstance(line['text'],str) and line['text'].strip() and '\n' not in line['text'] and '\r' not in line['text'],'BUTTON_LABEL_LINES_FIELDS')
        require(type(line['fontSize']) in (int,float) and math.isfinite(line['fontSize']) and line['fontSize']>0 and line['fontWeight'] in ('normal','bold') and line['align'] in ('left','center','right'),'BUTTON_LABEL_LINES_STYLE')
        r=line['layout'];require(isinstance(r,dict) and set(r)=={'x','y','width','height'} and all(type(v) in (int,float) and math.isfinite(v) for v in r.values()),'BUTTON_LABEL_LINES_BOUNDS')
        require(r['x']>=0 and r['y']>=0 and r['width']>0 and r['height']>=line['fontSize']*1.25 and (size is None or (r['x']+r['width']<=size[0] and r['y']+r['height']<=size[1])),'BUTTON_LABEL_LINES_BOUNDS')
        require(not any(r['x']<q['x']+q['width'] and r['x']+r['width']>q['x'] and r['y']<q['y']+q['height'] and r['y']+r['height']>q['y'] for q in rects),'BUTTON_LABEL_LINES_OVERLAP');rects.append(r)
    require(label is None or '\n'.join(l['text'] for l in lines)==label,'BUTTON_LABEL_LINES_TEXT_MISMATCH')
    return lines

File: game-ui-harnesses/ui-component-harness/src/button-label-lines.ts
/** Explicit target-component-local text lines; no inferred content or font fitting. */
export interface ButtonLabelLines {
  version: '1.0';
  coordinateSpace: 'target-component-local';
  lines: Array<{text:string; fontSize:number; fontWeight:'normal'|'bold'; align:'left'|'center'|'right'; layout:{x:number;y:number;width:number;height:number}}>;
}
export function buttonLabelLinesError(value:unknown,label:unknown,width:number,height:number):string|undefined {
  const object=(v:any,keys:string[])=>v&&typeof v==='object'&&!Array.isArray(v)&&Object.keys(v).length===keys.length&&keys.every(k=>Object.hasOwn(v,k));
  const v=value as ButtonLabelLines;
  if(!object(v,['version','coordinateSpace','lines'])||v.version!=='1.0'||v.coordinateSpace!=='target-component-local')return 'BUTTON_LABEL_LINES_VERSION';
  if(!Array.isArray(v.lines)||v.lines.length<1||v.lines.length>8)return 'BUTTON_LABEL_LINES_COUNT';
  for(const [i,line] of v.lines.entries()) {
    if(!object(line,['text','fontSize','fontWeight','align','layout'])||typeof line.text!=='string'||!line.text.trim()||/[\r\n]/.test(line.text))return 'BUTTON_LABEL_LINES_FIELDS';
    if(typeof line.fontSize!=='number'||!Number.isFinite(line.fontSize)||line.fontSize<=0||!['normal','bold'].includes(line.fontWeight)||!['left','center','right'].includes(line.align))return 'BUTTON_LABEL_LINES_STYLE';
    const r=line.layout;
    if(!object(r,['x','y','width','height'])||![r.x,r.y,r.width,r.height].every(n=>typeof n==='number'&&Number.isFinite(n))||r.x<0||r.y<0||r.width<=0||r.height<line.fontSize*1.25||r.x+r.width>width||r.y+r.height>height)return 'BUTTON_LABEL_LINES_BOUNDS';
    if(v.lines.slice(0,i).some(q=>r.x<q.layout.x+q.layout.width&&r.x+r.width>q.layout.x&&r.y<q.layout.y+q.layout.height&&r.y+r.height>q.layout.y))return 'BUTTON_LABEL_LINES_OVERLAP';
  }
  if(v.lines.map(l=>l.text).join('\n')!==label)return 'BUTTON_LABEL_LINES_TEXT_MISMATCH';
}
